### PR TITLE
Matching causatives found in research variants from other cases

### LINKED ISSUE #3512: Matching causatives missing
We have a case where Matching Causatives from other cases, is not visible on the first case page . But we found a variant that has been classified as causative in a earlier case, which is also evident from the variant page where the text "Matching causatives from other cases: xxx" is seen at the top of the page. 

Why is it missing on the first case page? Bug or related to the specific case/variant?  It seems to work on other cases we have (different variants). 
/Sofie

### PR DESCRIPTION
**Pull request by northwestwitch**
Fix #3512 

Current code performs a strict matching of variants _ids (created as md5 keys using chr, start, ref, alt, **clinical or research**), so when a case has clinical variants there won't be any matches for those variants in another case that has research variants marked as causative. 

I've changed the code to look for causatives also among research variants by using the 2 possible display names of a variant:
- chr_start_ref_alt_clinical
- chr_start_ref_alt_research

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the paxa software installed on `hasta`:
 - Log in into hasta: `ssh <USER.NAME>@hasta.scilifelab.se`
 - Activate the staging environment with the command `us` (Use Stage)
 - Run the command `paxa` and follow the instructions. Make sure to specify scout-stage as the resource you request and the server cg-vm1 as server.
1. Log out from the hasta server.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, log out from `cg-vm1` and log in again in the `hasta` server, repeat the `hasta` and `paxa` procedure, which will release the allocated resource (scout-stage) to be used for testing by other users.
</details>

**How to test, locally**:
1. Load the the demo case 643595: `scout --demo load case scout/demo/643595.config.yaml`
2. Load research variants for the demo case 643594: `scout --demo load research -c internal_id -i cust000 --force`
3. Go to case 643594, **research variants**, and mark as causative the usual POT1 var: http://localhost:5000/cust000/643594/80e1bdac78e83339117e4929f116a147
4. On main branch, if you go to the corresponding var of the close case 643595, it won't show any matching causatives
5. Switch to this branch and check that now the matching causative (research var on case 643594) is shown

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by DN 
- [x] tests executed by CR, DN

### CODE REVIEW

**Comment by northwestwitch** [comment]
This PR is blocked by #3510. Let's see what happens after fixing this issue

File: scout/adapter/mongo/variant.py [hunk 1]
```diff
@@ -524,17 +545,31 @@ def check_causatives(self, case_obj=None, institute_obj=None, limit_genes=None):
...
+
+            # Collect other causative variant's coords and change (exclude if it's clinical or research)
+            other_var_displ_name = var_event.get("subject").split("_")[
+                0:4
+            ]  # example: [ "17", "7577559", "G" "A"]
 
-            if other_causative_id in other_case.get("partial_causatives", {}).keys():
-                positional_variant_ids.add(var_event["variant_id"])
+            other_display_names.add("_".join(other_var_displ_name + ["clinical"]))
+            other_display_names.add("_".join(other_var_displ_name + ["research"]))
```
Comment (dnil) [inline scout/adapter/mongo/variant.py:565]: Wouldn't it be faster and easier to md5sum these into variant_ids and uses as positional_variant_ids?

(I'm still annoyed by the docs btw; we have sooo many places where it says variant_id is just positional id, nothing about clinical/research type. Still, this version, with type in addition to the simple id is correct according to tests.)

**Review by dnil — CHANGES_REQUESTED** [review]
This works, and solves the issue, but I'm afraid it would get slow with time, querying against unindexed fields on the variant collection. I suggest re-hashing the display names you generate in this PR to variant_ids to search indexed instead.

File: scout/adapter/mongo/variant.py (hunk 1 above)
Comment (northwestwitch) [reply scout/adapter/mongo/variant.py:565]: hehe that's how I did the PR in the beginning, but for some technical problem I've changed it. I'll check again!

Comment (northwestwitch) [reply scout/adapter/mongo/variant.py]: It wasn't that hard after all. I've changed the code to use md5 keys instead. Check it out when you have time!

File: scout/models/variant/variant.py [hunk 2]
```diff
@@ -3,12 +3,9 @@
...
     # document_id is a md5 string created by institute_genelist_caseid_variantid:
     _id=str,  # required, same as document_id
-    document_id=str,  # required
-    # variant_id is a md5 string created by chrom_pos_ref_alt (simple_id)
-    variant_id=str,  # required
-    # display name is variant_id (no md5)
-    display_name=str,  # required
-    # chrom_pos_ref_alt
+    document_id=str,  # required. Same as _id
+    variant_id=str,  # required. A md5 string created by chrom_pos_ref_alt_[clinical/research]
```
Comment (dnil) [inline scout/models/variant/variant.py:7]: 👍

File: scout/models/variant/variant.py [hunk 3]
```diff
@@ -3,12 +3,9 @@
...
+    display_name=str,  # required. no md5. chrom_pos_ref_alt_[clinical/research]
```
Comment (dnil) [inline scout/models/variant/variant.py:8]: Maybe add a comment for simple_id as well now?

File: scout/models/variant/variant.py (hunk 2 above)
Comment (dnil) [reply scout/models/variant/variant.py:7]: (you could call that `[c/r]` variant_type as well; defined below)

**Review by dnil — APPROVED** [review]
Thank you! This looks easier to maintain.

File: scout/models/variant/variant.py (hunk 2 above)
Comment (northwestwitch) [reply scout/models/variant/variant.py]: Done!

File: scout/models/variant/variant.py (hunk 3 above)
Comment (northwestwitch) [reply scout/models/variant/variant.py]: Done!

**Pull request merged**